scale-encode/src/impls/composite.rs

```diff
@@ -35,119 +97,168 @@ use scale_info::{PortableRegistry, TypeDef};
 /// }
 ///
 /// impl EncodeAsType for MyType {
-///     fn encode_as_type_to(&self, type_id: u32, types: &PortableRegistry, out: &mut Vec<u8>) -> Result<(), Error> {
-///         Composite([
-///             (Some("foo"), &self.foo as &dyn EncodeAsType),
-///             (Some("bar"), &self.bar as &dyn EncodeAsType),
-///             (Some("wibble"), &self.wibble as &dyn EncodeAsType)
-///         ].into_iter()).encode_as_type_to(type_id, types, out)
+///     fn encode_as_type_to<R: TypeResolver>(
+///         &self,
+///         type_id: &R::TypeId,
+///         types: &R,
+///         out: &mut Vec<u8>
+///     ) -> Result<(), Error> {
+///         Composite::new([
+///             (Some("foo"), CompositeField::new(&self.foo)),
+///             (Some("bar"), CompositeField::new(&self.bar)),
+///             (Some("wibble"), CompositeField::new(&self.wibble))
+///         ].into_iter()).encode_composite_as_type_to(type_id, types, out)
 ///     }
 /// }
 /// ```
-pub struct Composite<Vals>(pub Vals);
+///
+/// [`Composite`] cannot implement [`EncodeAsType`] itself, because it is tied to being
+/// encoded with a specific `R: TypeResolver`, whereas things implementing [`EncodeAsType`]
+/// need to be encodable using _any_ [`TypeResolver`]. This is ultimately because
+/// [`EncodeAsType`] is not object safe, which prevents it from being used to describe
+/// [`CompositeFields`][CompositeField].
+pub struct Composite<R, Vals> {
+    vals: Vals,
+    marker: core::marker::PhantomData<R>,
+}
 
-impl<'a, Vals> EncodeAsType for Composite<Vals>
+impl<'a, R, Vals> Composite<R, Vals>
 where
-    Vals: ExactSizeIterator<Item = (Option<&'a str>, &'a dyn EncodeAsType)> + Clone,
+    R: TypeResolver + 'a,
+    Vals: ExactSizeIterator<Item = (Option<&'a str>, CompositeField<'a, R>)> + Clone,
 {
-    fn encode_as_type_to(
+    /// Construct a new [`Composite`] type by providing an iterator over
+    /// the fields that it contains.
+    ///
+    /// ```rust
+    /// use scale_encode::{ Composite, CompositeField };
+    /// use scale_info::PortableRegistry;
+    ///
+    /// Composite::<PortableRegistry, _>::new([
+    ///     (Some("foo"), CompositeField::new(&123)),
+    ///     (Some("bar"), CompositeField::new(&"hello"))
+    /// ].into_iter());
+    /// ```
+    pub fn new(vals: Vals) -> Self {
+        Composite {
+            vals,
+            marker: core::marker::PhantomData,
+        }
+    }
+
+    /// A shortcut for [`Self::encode_composite_as_type_to()`] for when you just
```

Comment:
Little tricky to understand the docs but I think you want to say that this internally allocates a Vec and returns it.

Instead of passing a `&mut [u8]` or `&mut Vec<u8>` which `encode_composite_as_type_to` does

Reply:
Exactly! I'll tweak the docs and see if I can make them a bit clearer!